examples/jsm/controls/OrbitControls.js

```diff
@@ -271,6 +269,77 @@ class OrbitControls extends EventDispatcher {
 
 				}
 
+				// adjust camera position
+				let zoomChanged = false;
+				if ( scope.zoomToCursor ) {
+
+					let newRadius;
+					if ( scope.object.isPerspectiveCamera ) {
+
+						// move the camera down the pointer ray
+						// this method avoids floating point error
+						const prevRadius = offset.length();
+						newRadius = prevRadius * scale;
+						newRadius = Math.max( scope.minDistance, Math.min( scope.maxDistance, newRadius ) );
+
+						const radiusDelta = prevRadius - newRadius;
+						scope.object.position.addScaledVector( dollyDirection, radiusDelta );
+
+					} else if ( scope.object.isOrthographicCamera ) {
+
+						// adjust the ortho camera position based on zoom changes
+						const mouseBefore = new Vector3( mouse.x, mouse.y, 0 );
+						mouseBefore.unproject( scope.object );
+
+						scope.object.zoom = Math.max( scope.minZoom, Math.min( scope.maxZoom, scope.object.zoom / scale ) );
+						scope.object.updateProjectionMatrix();
+						zoomChanged = true;
+
+						const mouseAfter = new Vector3( mouse.x, mouse.y, 0 );
+						mouseAfter.unproject( scope.object );
+
+						scope.object.position.sub( mouseAfter ).add( mouseBefore );
+						newRadius = offset.length();
+
+					} else {
+
+						newRadius = offset.length();
+
+					}
+
+					scope.object.updateMatrixWorld();
+
+					if ( this.screenSpacePanning ) {
+
+						// position the orbit target in front of the new camera position
+						scope.target.set( 0, 0, - 1 )
+							.transformDirection( scope.object.matrix )
+							.multiplyScalar( newRadius )
+							.add( scope.object.position );
+
+					} else {
+
+						// get the ray and translation plane to compute target
+						const ray = new Ray();
+						ray.origin.copy( scope.object.position );
+						ray.direction.set( 0, 0, - 1 ).transformDirection( scope.object.matrix );
+
+						const plane = new Plane();
+						plane.setFromNormalAndCoplanarPoint( scope.object.up, scope.target );
+
+						ray.intersectPlane( plane, scope.target );
```

Comment:
The general approach is to move the camera such that the point under the cursor is visually consistent after zooming and the target is moved to accommodate this camera position movement.

However - when using non-screenspace panning and trying to zoom to cursor when the camera is looking parallel to the plane this portion of code fails because we're using the camera ray and plane to determine where to place the target. Any recommendations? Should we just disable zoom to cursor when the dot product of the plane normal and ray  is under some threshold? Then we can recommend that users limit the angles in these cases.

Reply:
I think we should make it so that when you're parallel and zoom "towards the sky", we pick an arbitrary distance away and use that as if we had collided with the terrain, then use the same code. I.e. dolly the camera in that direction

Reply:
Unfortunately this won't work since the camera always has to face the target point. This means the camera will move up into the sky and then arc down towards the view point since the target has to stay on that plane - which isn't necessarily what someone would expect.

Zoom to cursor with map controls only really makes sense when limiting the camera control angles, imo, but we should have some approach that doesn't break here.

Reply:
btw,
> since the camera always has to face the target point

how can you possibly maintain both target and the point under cursor constraints? one of them will fail as soon as you move persp. camera in any way other than rotating around the axis formed by those 2 points.